src/bindings/generator/src/dbus_typing.py

```diff
@@ -0,0 +1,148 @@
+# SPDX-License-Identifier: GPL-2.0-or-later
+
+from typing import Tuple
+import dasbus.typing as DBusTypes
+
+basic_types = ["y", "b", "n", "q", "i", "u", "x", "t", "d", "s", "o", "g", "h", "v"]
+
+
+def map_basic_type_to_pytype(t: str) -> str:
+    if t == "y":
+        return DBusTypes.get_type_name(DBusTypes.Byte)
+    elif t == "b":
+        return DBusTypes.get_type_name(DBusTypes.Bool)
+    elif t == "n":
+        return DBusTypes.get_type_name(DBusTypes.Int16)
+    elif t == "q":
+        return DBusTypes.get_type_name(DBusTypes.UInt16)
+    elif t == "i":
+        return DBusTypes.get_type_name(DBusTypes.Int32)
+    elif t == "u":
+        return DBusTypes.get_type_name(DBusTypes.UInt32)
+    elif t == "x":
+        return DBusTypes.get_type_name(DBusTypes.Int16)
+    elif t == "t":
+        return DBusTypes.get_type_name(DBusTypes.UInt64)
+    elif t == "d":
+        return DBusTypes.get_type_name(DBusTypes.Double)
+    elif t == "s":
+        return DBusTypes.get_type_name(DBusTypes.Str)
+    elif t == "o":
+        return DBusTypes.get_type_name(DBusTypes.ObjPath)
+    elif t == "g":
+        return DBusTypes.get_type_name(DBusTypes.Str)
+    elif t == "h":
+        return DBusTypes.get_type_name(DBusTypes.UnixFD)
+    elif t == "v":
+        return DBusTypes.get_type_name(DBusTypes.Variant)
+
+    raise Exception(f"Unknown basic type '{t}'")
+
+
+def map_array_type_to_pytype(t: str) -> Tuple[str, int]:
+    if len(t) < 1 or t[0] != "a":
+        raise Exception(
+            "Invalid type definition for array detected in '{t}' - No 'a' found!"
+        )
+
+    if t[1] == "y":
+        return "bytes", 2
+    elif t[1] in basic_types:
+        return f"List[{map_basic_type_to_pytype(t[1])}]", 2
+    elif t[1] == "(":
+        mapped, num_read = map_struct_to_pytype(t[1:])
+        return f"List[{mapped}]", num_read + 3
+    elif t[1] == "{":
+        if t[2] not in basic_types:
+            raise Exception(f"Invalid dict definition for '{t}' - key must be a basic type")
+        key = map_basic_type_to_pytype(t[2])
+
+        if t[3] in basic_types and t[4] == "}":
+            value = map_basic_type_to_pytype(t[3])
+            ret = f"Dict[{key}, {value}]"
+            if ret == "Dict[str, Variant]":
+                ret = "Structure"
+            return ret, 5
+        if t[3] == "a":
+            mapped, num_read = map_array_type_to_pytype(t[3:])
+            return f"Dict[{key}, {mapped}]", num_read + 4
+        if t[3] == "(":
+            mapped, num_read = map_struct_to_pytype(t[3:])
+            return f"Dict[{key}, {mapped}]", num_read + 6
```

Comment:
Can we have some constants to these magic numbers like '+ 6', '+ 4', '+ 3' ? I am sure you understand very well these numbers but in the future for maintenance would be easier.

Reply:
Its a bit tricky with the naming here. Basically, the `num_read` is the number of characters read - which helps to only iterate over the dbus type string once (`O(n)`). In the marked example for the +6 its due to the dict with a struct as value, e.g. `a{s(is)}` - so it adds the `a`, `{`, `s`, `(`, `)`, `}` to the count. I did a bit of renaming and added inline comments to explain whats going on. 
Not the best approach, but this was hacked down pretty fast - we can refactor this in the future :)